main/src/ca/uwaterloo/flix/language/phase/unification/EffUnification2.scala

```diff
@@ -0,0 +1,146 @@
+    */
+  private def toEquation(p: (Type, Type, SourceLocation))(implicit renv: RigidityEnv, m: Bimap[Type.Var, Int]): Equation = {
+    val (tpe1, tpe2, loc) = p
+    Equation.mk(toTerm(tpe1), toTerm(tpe2), loc)
+  }
+
+  /**
+    * Returns the given type `t` as term.
+    *
+    * Uses the given bimap `m` to map type variables to term variables.
+    *
+    * The rigidity environment `renv` is used to map rigid type variables to constants and flexible type variables to term variables.
+    */
+  private def toTerm(t: Type)(implicit renv: RigidityEnv, m: Bimap[Type.Var, Int]): Term = Type.eraseTopAliases(t) match {
+    case Type.Pure => Term.True
+    case Type.Univ => Term.False
+
+    case t: Type.Var => m.getForward(t) match {
+      case None => throw InternalCompilerException(s"Unexpected unbound type variable: '$t'.", t.loc)
+      case Some(x) => renv.get(t.sym) match {
+        case Rigidity.Flexible => Term.Var(x) // A flexible variable is a real variable.
+        case Rigidity.Rigid => Term.Cst(x) // A rigid variable is a constant.
+      }
+    }
+
+    case Type.Apply(Type.Cst(TypeConstructor.Complement, _), tpe1, _) => Term.mkNot(toTerm(tpe1))
+    case Type.Apply(Type.Apply(Type.Cst(TypeConstructor.Union, _), tpe1, _), tpe2, _) => Term.mkAnd(toTerm(tpe1), toTerm(tpe2))
+    case Type.Apply(Type.Apply(Type.Cst(TypeConstructor.Intersection, _), tpe1, _), tpe2, _) => Term.mkOr(toTerm(tpe1), toTerm(tpe2))
+
+    case _ => throw InternalCompilerException(s"Unexpected type: '$t'.", t.loc)
+  }
+
+  /**
+    * Returns a Boolean substitution as a regular type substitution.
+    */
+  private def fromBoolSubst(s: FastBoolUnification.BoolSubstitution)(implicit m: Bimap[Type.Var, Int]): Substitution = {
+    Substitution(s.m.foldLeft(Map.empty[Symbol.KindedTypeVarSym, Type]) {
+      case (macc, (k, v)) => macc + (m.getBackward(k).get.sym -> fromTerm(v, SourceLocation.Unknown))
+    })
+  }
+
+  /**
+    * Returns the given term `t` as a type.
+    *
+    * Uses the given bimap `m` to map term variables to type variables.
+    * Uses the given source location `loc` as the source location for all sub-terms.
+    *
+    * Both constants and variables are mapped back to type variables. The rigidity environment, in the type world,
+    * distinguishes their rigidity or flexibility.
+    */
+  private def fromTerm(t: Term, loc: SourceLocation)(implicit m: Bimap[Type.Var, Int]): Type = t match {
+    case Term.True => Type.Pure
+    case Term.False => Type.Univ
+    case Term.Cst(c) => m.getBackward(c).get // Safe: We never introduce new variables.
+    case Term.Var(x) => m.getBackward(x).get // Safe: We never introduce new variables.
+    case Term.Not(t) => Type.mkComplement(fromTerm(t, loc), loc)
+    case Term.And(csts, vars, rest) =>
+      val ts = csts.toList.map(fromTerm(_, loc)) ++ vars.toList.map(fromTerm(_, loc)) ++ rest.map(fromTerm(_, loc))
+      Type.mkIntersection(ts, loc)
```

Comment:
True -> Pure and False -> Univ is consistent though

Reply:
Its a good point. I suggest that this will not really be resolved until we actually have sets. 

I can of course swap them, but then I have to refactor a lot of code (for little gain). I would just leave it as is for now.

Reply:
Btw. the unification problems I was given uses intersection, so I am still not sure what is going on...

Reply:
But you should at least be consistent in the same file

Reply:
Right. I think I fixed it now.